<a href="https://colab.research.google.com/github/nxxk23/PROJECT/blob/main/WideTable.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import pandas as pd

hemo = pd.read_csv('/content/drive/MyDrive/Thesis Project Kidney/Hemodialysis.csv', encoding='latin1')
# Create a list from hemo['PatientUID']
patient_uid_list = hemo['PatientUID'].unique().tolist()

In [3]:
import pandas as pd
vital = pd.read_csv('/content/drive/MyDrive/Thesis Project Kidney/vitalsign.csv')
# Remove patient UIDs from visit['PatientUID'] that are not in the list
vital = vital[~vital['PatientUID'].isin(patient_uid_list)]
vital['PatientUID'].nunique()

20714

In [4]:
visit = pd.read_csv('/content/drive/MyDrive/Thesis Project Kidney/Visit.csv')

# Remove patient UIDs from visit['PatientUID'] that are not in the list
visit = visit[~visit['PatientUID'].isin(patient_uid_list)]
visit['PatientUID'].nunique()

<ipython-input-4-3ec828bcbf86>:1: DtypeWarning: Columns (8) have mixed types. Specify dtype option on import or set low_memory=False.
  visit = pd.read_csv('/content/drive/MyDrive/Thesis Project Kidney/Visit.csv')


20967

In [5]:
import pandas as pd
lab = pd.read_csv('/content/drive/MyDrive/Thesis Project Kidney/Lab.csv')
# Remove patient UIDs from visit['PatientUID'] that are not in the list
lab = lab[~lab['PatientUID'].isin(patient_uid_list)]
lab['PatientUID'].nunique()

20967

In [6]:
slope = pd.read_excel("/content/drive/MyDrive/Thesis Project Kidney/Assign/Assign 3: Re-CountSlope & Quartile/LabSlope.xlsx")
slope['PatientUID'].nunique()

17711

In [7]:
diagnosis = pd.read_csv('/content/drive/MyDrive/Thesis Project Kidney/Diagnosis.csv')
# Remove patient UIDs from visit['PatientUID'] that are not in the list
diagnosis = diagnosis[~diagnosis['PatientUID'].isin(patient_uid_list)]
diagnosis['PatientUID'].nunique()

20719

## 🗒**Note**
> **จำนวนผู้ป่วยในแต่ละไฟล์ (หลังตัด Hemo)**
- vital == 20714
- visit == 20976
- lab  == 20967
- diag == 20719
- slope == 17711

### **Test Vital**

In [8]:
vital['ResultItemName'].unique()

array(['Weight', 'BMI', 'Pulse Rate', 'Respiratory Rate', 'Systolic BP',
       'BSA', 'Diastolic BP', 'Height', 'Temperature', 'O2 Sat',
       'Pain Score', 'MeanPressure', 'Waistline', 'IrregularHeartBeat',
       'AverageMeasurement', '6MWT', 'Head Circumference', 'Fat'],
      dtype=object)

In [9]:
vital.columns

Index(['LabNo', 'PatientUID', 'PatientVisitUID', 'OrderItemName',
       'ResultItemName', 'ResultValue', 'CWhen'],
      dtype='object')

In [10]:
import pandas as pd

# Assuming 'vital' is a DataFrame containing your data
# Columns: ['LabNo', 'PatientUID', 'PatientVisitUID', 'OrderItemName', 'ResultItemName', 'ResultValue', 'CWhen']

# Use pivot_table if there are duplicate entries for the same 'ResultItemName' and 'ResultValue' combination
pivot_table = vital.pivot_table(index=['LabNo', 'PatientUID', 'PatientVisitUID', 'OrderItemName'], columns='ResultItemName', values='ResultValue', aggfunc='first')

# Use pivot if there are no duplicate entries for the same 'ResultItemName' and 'ResultValue' combination
pivot = vital.pivot(index=['LabNo', 'PatientUID', 'PatientVisitUID', 'OrderItemName'], columns='ResultItemName', values='ResultValue')


# Now, you have a wide-format DataFrame where 'ResultItemName' values become columns

In [11]:
pivot

ResultItemName                                      6MWT AverageMeasurement  \
LabNo PatientUID PatientVisitUID OrderItemName                                
1     2299       0               Vital Signs Record  NaN                NaN   
                 2024049         Vital Signs Record  NaN                NaN   
                 2237669         Vital Signs Record  NaN                NaN   
                 2396155         Vital Signs Record  NaN                NaN   
                 2480434         Vital Signs Record  NaN                NaN   
...                                                  ...                ...   
1326  1472363    9805481         Vital Signs Record  NaN                NaN   
1327  1472363    9805481         Vital Signs Record  NaN                NaN   
1328  1472363    9805481         Vital Signs Record  NaN                NaN   
1329  1472363    9805481         Vital Signs Record  NaN                NaN   
1330  1472363    9805481         Vital Signs Record  NaN                NaN   

ResultItemName                                         BMI   BSA Diastolic BP  \
LabNo PatientUID PatientVisitUID OrderItemName                                  
1     2299       0               Vital Signs Record  20.73  1.36           79   
                 2024049         Vital Signs Record  20.17  1.39           65   
                 2237669         Vital Signs Record  20.36  1.38           83   
                 2396155         Vital Signs Record     20  1.37           67   
                 2480434         Vital Signs Record  18.51  1.31           96   
...                                                    ...   ...          ...   
1326  1472363    9805481         Vital Signs Record    NaN   NaN          NaN   
1327  1472363    9805481         Vital Signs Record    NaN   NaN          NaN   
1328  1472363    9805481         Vital Signs Record    NaN   NaN          NaN   
1329  1472363    9805481         Vital Signs Record    NaN   NaN          NaN   
1330  1472363    9805481         Vital Signs Record    NaN   NaN          NaN   

ResultItemName                                       Fat Head Circumference  \
LabNo PatientUID PatientVisitUID OrderItemName                                
1     2299       0               Vital Signs Record  NaN                NaN   
                 2024049         Vital Signs Record  NaN                NaN   
                 2237669         Vital Signs Record  NaN                NaN   
                 2396155         Vital Signs Record  NaN                NaN   
                 2480434         Vital Signs Record  NaN                NaN   
...                                                  ...                ...   
1326  1472363    9805481         Vital Signs Record  NaN                NaN   
1327  1472363    9805481         Vital Signs Record  NaN                NaN   
1328  1472363    9805481         Vital Signs Record  NaN                NaN   
1329  1472363    9805481         Vital Signs Record  NaN                NaN   
1330  1472363    9805481         Vital Signs Record  NaN                NaN   

ResultItemName                                      Height IrregularHeartBeat  \
LabNo PatientUID PatientVisitUID OrderItemName                                  
1     2299       0               Vital Signs Record    148                NaN   
                 2024049         Vital Signs Record    151                NaN   
                 2237669         Vital Signs Record    150                NaN   
                 2396155         Vital Signs Record    150                NaN   
                 2480434         Vital Signs Record    149                NaN   
...                                                    ...                ...   
1326  1472363    9805481         Vital Signs Record    NaN                NaN   
1327  1472363    9805481         Vital Signs Record    NaN                NaN   
1328  1472363    9805481         Vital Signs Record    NaN                NaN   

In [12]:
# Reset the index if you want to move the grouped columns back to regular columns
#pivot_table_reset = pivot_table.reset_index()
#pivot_reset = pivot.reset_index()

### **Test Lab**

In [13]:
lab

LabNo  PatientUID  PatientVisitUID  \
0             1        2299          2024049   
1             1        2299          2024049   
2             1        2299          2024049   
3             1        2299          2024049   
4             1        2299          2024049   
...         ...         ...              ...   
21641153    601      822860          5140446   
21641156    602      822860          5140446   
21641159    603      822860          5140446   
21641162    604      822860          5140446   
21641165    605      822860          5140446   

                                           Location   RequestItemName  \
0                     เวชศาสตร์นิวเคลียร์ (OPD)-D15        BMD LUMBAR   
1         ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร    ALBUMIN, 32403   
2         ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร     Alkaline Phos   
3         ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร        ALT, 32311   
4         ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร        AST, 32310   
...                                             ...               ...   
21641153  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร  Dextrostix(POCT)   
21641156  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร  Dextrostix(POCT)   
21641159  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร  Dextrostix(POCT)   
21641162  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร  Dextrostix(POCT)   
21641165  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร  Dextrostix(POCT)   

            ResultItemName                    CWhen ResultDttm ResultValue  \
0         Formatted Result  2558-02-19 16:16:51.423        NaN          FT   
1           ALBUMIN, 32403  2558-02-19 10:32:20.907        NaN         4.1   
2               ALP, 32309  2558-02-19 10:32:21.610        NaN          44   
3               ALT, 32311  2558-02-19 10:32:21.390        NaN          17   
4               AST, 32310  2558-02-19 10:32:21.500        NaN          23   
...                    ...                      ...        ...         ...   
21641153         DTX(POCT)  2561-10-19 11:00:42.207        NaN         228   
21641156         DTX(POCT)  2561-10-19 17:24:45.417        NaN         240   
21641159         DTX(POCT)  2561-10-19 22:40:42.073        NaN         171   
21641162         DTX(POCT)  2561-10-20 05:39:12.677        NaN         191   
21641165         DTX(POCT)  2561-10-20 10:53:23.537        NaN         308   

           Unit ReferenceRange  
0           NaN            NaN  
1           NaN      3.8 - 5.4  
2           NaN       42 - 121  
3           NaN         4 - 36  
4           NaN        12 - 32  
...         ...            ...  
21641153  mg/dL              -  
21641156  mg/dL              -  
21641159  mg/dL              -  
21641162  mg/dL              -  
21641165  mg/dL              -  

[17728574 rows x 11 columns]

In [14]:
import pandas as pd

# Replace the link with your Google Sheets CSV link for the specific sheet
sheet_link = "https://docs.google.com/spreadsheets/d/1b3VqChmx_Ezb5BCC97Pe3EEmCKFubSIQCsSBre5z8_M/gviz/tq?tqx=out:csv&sheet=Lab"

# Read the CSV file directly into a Pandas DataFrame
df = pd.read_csv(sheet_link)

# Now you have your data in a Pandas DataFrame
df

Unnamed: 0                                   Location  \
0              0  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร   
1              1  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร   
2              2  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร   
3              3  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร   
4              4  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร   
...          ...                                        ...   
2985        2985       ห้องปฏิบัติการภูมิคุ้มกันวิทยาคลินิก   
2986        2986       ห้องปฏิบัติการภูมิคุ้มกันวิทยาคลินิก   
2987        2987       ห้องปฏิบัติการภูมิคุ้มกันวิทยาคลินิก   
2988        2988       ห้องปฏิบัติการภูมิคุ้มกันวิทยาคลินิก   
2989        2989       ห้องปฏิบัติการภูมิคุ้มกันวิทยาคลินิก   

              ResultItemName             RequestItemName   Count  
0          CREATININE, 32202                  CREATININE  464626  
1     eGFR(CKD-EPI) age >=18                  CREATININE  435208  
2           POTASSIUM, 32103                   POTASSIUM  399456  
3                 BUN, 32201                         BUN  388865  
4         BICARBONATE, 32105                 BICARBONATE  386526  
...                      ...                         ...     ...  
2985         MUMP IGM, 36650                    MUMP IGM       1  
2986         MUMP IGM, 36650             MUMP IGM, 36650       1  
2987           Microsomal Ab               Microsomal Ab       1  
2988                Mosquito  i71 Mosquito(Specific IgE)       1  
2989             FSH 120 min                 FSH 120 min       1  

[2990 rows x 5 columns]

In [15]:
a = df[df['Count']>=100000]
a

Unnamed: 0                                   Location  \
0            0  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร   
1            1  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร   
2            2  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร   
3            3  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร   
4            4  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร   
..         ...                                        ...   
56          56  ห้องปฏิบัติการเคมีคลินิก เวชศาสตร์ชันสูตร   
57          57            ห้องปฏิบัติการจุลทรรศน์วินิจฉัย   
58          58            ห้องปฏิบัติการจุลทรรศน์วินิจฉัย   
59          59            ห้องปฏิบัติการจุลทรรศน์วินิจฉัย   
60          60            ห้องปฏิบัติการจุลทรรศน์วินิจฉัย   

             ResultItemName       RequestItemName   Count  
0         CREATININE, 32202            CREATININE  464626  
1    eGFR(CKD-EPI) age >=18            CREATININE  435208  
2          POTASSIUM, 32103             POTASSIUM  399456  
3                BUN, 32201                   BUN  388865  
4        BICARBONATE, 32105           BICARBONATE  386526  
..                      ...                   ...     ...  
56  DIRECT BILIRUBIN, 32207      DIRECT BILIRUBIN  128979  
57              Slide No.    Complete blood count  125066  
58                 SQ-Epi              Urinalysis  119335  
59                     PT                  PT/INR  101035  
60                    INR                  PT/INR  100291  

[61 rows x 5 columns]

In [16]:
a['RequestItemName'].unique()

array(['CREATININE', 'POTASSIUM', 'BUN', 'BICARBONATE', 'SODIUM',
       'CHLORIDE', 'Dextrostix(POCT)', 'Complete blood count', 'ALBUMIN',
       'CHOLESTEROL', 'ALT', 'AST', 'Urinalysis', 'Alkaline Phos',
       'LDL-CHOL', 'SUGAR IN NAF-PLASMA', 'CALCIUM', 'PHOSPHORUS',
       'HB A1C', 'TRIGLYCERIDE', 'HDL-C', 'MAGNESIUM', 'Globulin',
       'TOTAL PROTEIN', 'TOTAL BILIRUBIN', 'DIRECT BILIRUBIN', 'PT/INR'],
      dtype=object)

In [17]:
import pandas as pd

# Assuming 'lab' is a DataFrame containing your data
# Columns: ['LabNo', 'PatientUID', 'PatientVisitUID', 'OrderItemName', 'ResultItemName', 'ResultValue', 'CWhen']

# Select the items with more than 100000 rows
selected_items = ['CREATININE', 'POTASSIUM', 'BUN', 'BICARBONATE', 'SODIUM',
                  'CHLORIDE', 'Dextrostix(POCT)', 'Complete blood count', 'ALBUMIN',
                  'CHOLESTEROL', 'ALT', 'AST', 'Urinalysis', 'Alkaline Phos',
                  'LDL-CHOL', 'SUGAR IN NAF-PLASMA', 'CALCIUM', 'PHOSPHORUS',
                  'HB A1C', 'TRIGLYCERIDE', 'HDL-C', 'MAGNESIUM', 'Globulin',
                  'TOTAL PROTEIN', 'TOTAL BILIRUBIN', 'DIRECT BILIRUBIN', 'PT/INR']

# Filter the DataFrame based on the selected items
filtered_lab = lab[lab['RequestItemName'].isin(selected_items)]

# Use pivot_table to pivot the DataFrame
pivot_table1 = filtered_lab.pivot_table(index=['LabNo', 'PatientUID', 'PatientVisitUID'], columns='RequestItemName', values='ResultValue', aggfunc='first')
pivot_table1


RequestItemName                  ALBUMIN  ALT  AST Alkaline Phos BICARBONATE  \
LabNo PatientUID PatientVisitUID                                               
1     2299       2024049             NaN  NaN  NaN            44         NaN   
                 2381054             NaN  NaN  NaN            47         NaN   
                 2499619             NaN  NaN  NaN           NaN         NaN   
                 2534514             NaN  NaN  NaN            50         NaN   
                 2706070             NaN  NaN  NaN           NaN         NaN   
...                                  ...  ...  ...           ...         ...   
601   822860     5140446             NaN  NaN  NaN           NaN         NaN   
602   822860     5140446             NaN  NaN  NaN           NaN         NaN   
603   822860     5140446             NaN  NaN  NaN           NaN         NaN   
604   822860     5140446             NaN  NaN  NaN           NaN         NaN   
605   822860     5140446             NaN  NaN  NaN           NaN         NaN   

RequestItemName                   BUN CALCIUM CHLORIDE CHOLESTEROL CREATININE  \
LabNo PatientUID PatientVisitUID                                                
1     2299       2024049          NaN     NaN      NaN         NaN        NaN   
                 2381054          NaN     NaN      NaN         NaN        NaN   
                 2499619          NaN     NaN      NaN         NaN        NaN   
                 2534514          NaN     NaN      NaN         NaN        NaN   
                 2706070          NaN     NaN      NaN         195        NaN   
...                               ...     ...      ...         ...        ...   
601   822860     5140446          NaN     NaN      NaN         NaN        NaN   
602   822860     5140446          NaN     NaN      NaN         NaN        NaN   
603   822860     5140446          NaN     NaN      NaN         NaN        NaN   
604   822860     5140446          NaN     NaN      NaN         NaN        NaN   
605   822860     5140446          NaN     NaN      NaN         NaN        NaN   

RequestItemName                   ... MAGNESIUM PHOSPHORUS POTASSIUM PT/INR  \
LabNo PatientUID PatientVisitUID  ...                                         
1     2299       2024049          ...       NaN        NaN       NaN    NaN   
                 2381054          ...       NaN        NaN       NaN    NaN   
                 2499619          ...       NaN        NaN       NaN    NaN   
                 2534514          ...       NaN        NaN       NaN    NaN   
                 2706070          ...       NaN        NaN       NaN    NaN   
...                               ...       ...        ...       ...    ...   
601   822860     5140446          ...       NaN        NaN       NaN    NaN   
602   822860     5140446          ...       NaN        NaN       NaN    NaN   
603   822860     5140446          ...       NaN        NaN       NaN    NaN   
604   822860     5140446          ...       NaN        NaN       NaN    NaN   
605   822860     5140446          ...       NaN        NaN       NaN    NaN   

RequestItemName                  SODIUM SUGAR IN NAF-PLASMA TOTAL BILIRUBIN  \
LabNo PatientUID PatientVisitUID                                              
1     2299       2024049            NaN                 NaN             NaN   
                 2381054            NaN                 NaN             NaN   
                 2499619            NaN                 NaN             NaN   
                 2534514            NaN                 NaN             NaN   
                 2706070            NaN                 NaN             NaN   
...                                 ...                 ...             ...   
601   822860     5140446            NaN                 NaN             NaN   
602   822860     5140446            NaN                 NaN             NaN   
603   822860     5140446            NaN                 NaN             NaN   
604   82